### ISSUE TITLE
Restore having only backup

### ISSUE THREAD

**Issue by vvitad** [pgbackrest/pgbackrest#2039]
1. pgBackRest 2.43
2. PostgreSQL 9.6:
3. Oracle Linux 7
4. source

Hello. I haven't been able to find any info on restoring if you have no previous stanza, no previous server, nada.
I have a clean server and contents of the repository, where there is 1 full and 1 incr comressed backups. 1 backup set to be exact

I install PostgreSQL, install pgBackrest. 
Then i've tried:
made pgbackrest.conf, where repo1-path points to the empty directory , created stanza
```
[global]
repo1-path=/var/lib/pgsql/backup
log-level-file=debug
process-max=2

[test3]
pg1-path=/var/lib/pgsql/9.6/data
pg1-port=5436

[global:archive-push]
compress-level=6

```

Then I change repo1-path to- the directory where my backup set lies and try restoring, it fails:
```
2023-03-29 13:38:14.379 P00   WARN: repo1: [FormatError] unable to load info file '/var/lib/pgsql/pgbackrest_backup/backup/test3/backup.info' or '/var/lib/pgsql/pgbackrest_backup/backup/test3/backup.info.copy':
                                    FormatError: key/value found outside of section at line 1: Salted__<9E><E5>b:<E0>^]&R<A6>i<DE><C1>+3^Q<EC>˔<9F>^?<F6>[&<83><C7><D2>
                                    FormatError: key/value found outside of section at line 1: Salted__<9E><E5>b:<E0>^]&R<A6>i<DE><C1>+3^Q<EC>˔<9F>^?<F6>[&<83><C7><D2>
                                    HINT: backup.info cannot be opened and is required to perform a backup.
                                    HINT: has a stanza-create been performed?
2023-03-29 13:38:14.379 P00  DEBUG:     command/exit::exitSafe: (result: 0, error: true, signalType: 0)
2023-03-29 13:38:14.379 P00  ERROR: [075]: no backup set found to restore
                                    --------------------------------------------------------------------
                                    If SUBMITTING AN ISSUE please provide the following information:

                                    version: 2.43
                                    command: restore
                                    options: --exec-id=20773-0409ad51 --log-level-console=info --log-level-file=debug --pg1-path=/var/lib/pgsql/9.6/data --process-max=2 --repo1-path=/var/lib/pgsql/pgbackrest_backup --stanza=test3

                                    stack trace:
                                    command/restore/restore.c:restoreBackupSet:378:(void)
                                    command/restore/restore.c:cmdRestore:(void)
                                    main.c:main:(debug log level required for parameters)
                                    --------------------------------------------------------------------

```

I am pretty sure I am doing something silly, could you please tell me, how do I restore in this situation?
I don't believe  that backup set is useless

**Comment by vvitad** [comment]
I have read this issue https://github.com/pgbackrest/pgbackrest/issues/1971, seemed similar
I rm pgdata contents, run scp -r user1@host-where-repo-lies/home/user1/pgbackrest/* /var/lib/pgsql/backup

```

-bash-4.2$ pgbackrest  --log-level-console=info --stanza=test3  --repo1-path=/var/lib/pgsql/pgbackrest_backup/ --archive-mode=off restore --pg1-path=/var/lib/pgsql/test3
2023-03-29 19:05:29.555 P00   INFO: restore command begin 2.43: --archive-mode=off --exec-id=8123-04124833 --log-level-console=info --log-level-file=debug --pg1-path=/var/lib/pgsql/test3 --process-max=2 --repo1-path=/var/lib/pgsql/pgbackrest_backup/ --stanza=test3
WARN: repo1: [FileMissingError] unable to load info file '/var/lib/pgsql/pgbackrest_backup/backup/test3/backup.info' or '/var/lib/pgsql/pgbackrest_backup/backup/test3/backup.info.copy':
      FileMissingError: unable to open missing file '/var/lib/pgsql/pgbackrest_backup/backup/test3/backup.info' for read
      FileMissingError: unable to open missing file '/var/lib/pgsql/pgbackrest_backup/backup/test3/backup.info.copy' for read
      HINT: backup.info cannot be opened and is required to perform a backup.
      HINT: has a stanza-create been performed?
ERROR: [075]: no backup set found to restore
2023-03-29 19:05:29.557 P00   INFO: restore command end: aborted with exception [075]

```


Were you saing I should rsync, not scp?

**Comment by pgstef** [comment]
Hi,

> if you have no previous stanza, no previous server, nada.

What do you want to restore if you don't have anything? :thinking: 
A "stanza" is just a grouping of configuration options and physically a file hierarchy in the repository.

I guess your implicit question here is how you can restore a backup from a "duplicated" repository. As long as the original repository isn't corrupted (contains all the manifest and info files), that could work.
What you'd need to do after the copy is to adjust your configuration to point to that copied repo, using some of the original configuration settings (like compression / encryption /...).

`scp -r user1@host-where-repo-lies/home/user1/pgbackrest/* /var/lib/pgsql/backup` and then you're using `--repo1-path=/var/lib/pgsql/pgbackrest_backup`.

So the error message is pretty clear: `/var/lib/pgsql/pgbackrest_backup/backup/test3/backup.info`. Does this file exists on disk?

Anyway, if you want help, you need to be more precise about the original configuration used on the main repo and what are the steps/commands you used to duplicate it and try a restore.

Regards,

**Comment by vvitad** [comment]
@pgstef , I've been waiting for you :) i pointed out to having no-stanza, because half of my attemtpts i've had a hint: has stanza been created?
>>Does this file exists on disk?
Of course it exists, I checked like a thousand times and even tried to open it, but it's binary. I still not sure what caused that error.

You are right though, my last comment doesn't make any sence. "Salted" eror was cause of I thought there was no cypher
FormatError: key/value found outside of section at line 1: Salted_
Once I added it , it has suscecfully restored. Couldn't start PG because it couldn't find checkpoints, so I had to reset it pg_xlogreset. 

I am goint to recreate the problem with checkpoints in the next coment.

**Comment by pgstef** [comment]
> hint: has stanza been created?

That's simply because it can't find the file in the repo your configured.

> so I had to reset it pg_xlogreset.

Don't do that. It will corrupt your cluster. It means you don't have the archives needed for the restore (or that the restore_command isn't set or not working correctly). That's hard to tell with so little info and no PG recovery logs.

Did you copied the archives directory too? (another option would be to add the `archive-copy=y` option into your backup host configuation so wals needed for the backup consistency get copied into the backup directory directly at the backup time.

**Comment by pgstef** [comment]
> "Salted" eror was cause of I thought there was no cypher
> FormatError: key/value found outside of section at line 1: Salted_

That's why we can't really help/give accurate advice without the proper information. In this case, for example, the settings used on the host where the original backup was taken.

**Comment by vvitad** [comment]
Also I had to delete backup.label file, cause It couldn't start. (Logs suggested deletion)

> Did you copied the archives directory too? (another option would be to add the `archive-copy=y` option into your backup host configuation so wals needed for the backup consistency get copied into the backup directory directly at the backup time.
Aren't they are in the /repo/archive?

I remember reading this and thinking 'hmm well i definitely won't need it'
`This slightly paranoid option protects against corruption in the WAL segment archive by storing the WAL segments required for consistency directly in the backup. WAL segments are still stored in the archive so this option will use additional space.
`


I have level 4 compression set for backup. If I set archive-copy=y, do I need to set archive-push setting, how should it look like? 

> It is best if the archive-push and backup commands have the same compress-type (e.g. lz4) when using this option. Otherwise, the WAL segments will need to be recompressed with the compress-type used by the backup, which can be fairly expensive depending on how much WAL was generated during the backup.

**Comment by pgstef** [comment]
> Also I had to delete backup.label file, cause It couldn't start. (Logs suggested deletion)

You totally shouldn't do that. That's only when you're not restoring from a backup (as the PG log should have stated).
That's exactly why the exclusive backup mode has been removed in PG 15. See [this](https://www.cybertec-postgresql.com/en/exclusive-backup-deprecated-what-now/) blog post that sums it up pretty clearly imo.

With the very clear conclusion:

> There is one last thing you have to do: you have to store the backup_label file with the checkpoint information along with the backup. The file must be present after the backup has been restored. Remember that if you start PostgreSQL on a restored data directory without the correct backup_label file, the result will be data corruption. This is because the pg_control file in the backup usually contains a later checkpoint than the one taken during pg_start_backup.

> I remember reading this and thinking 'hmm well i definitely won't need it'

Well you wouldn't need it for production only. If you copy the wal archives too in your duplicated repo, PG might be able to use them for the recovery consistency. But apparently you're not handling the recovery very properly. I gave a [talk](https://pgstef.github.io/talks/en/20230323_pgDayParis_PostgreSQL-Recovery-Targets.reveal.pdf) last week about that specific topic. But since (once again) you're not very precise/accurate regarding your configs, commands used and logs, I probably won't be able to help you more on this.

**Comment by pgstef** [comment]
I don't think `repo1-path=/var/lib/pgbackrest` is needed in your primary config.

> Then I mounted repo to the test server and run a restore

> So i chmod 777 -R on this directory

Did you mounted the production repo or copied/rsynced it?
Chmod on your production repo isn't a good idea imo. What you could do is adjust your mount options to map it to the good user:group and only give it RO access.

`sudo pgbackrest  --log-level-console=info --stanza=test3  --repo1-path=/mnt/pgbackrest/  restore --pg1-path=/var/lib/pgsql/test3`

Why sudo? The pgbackrest commands should usually run under your PG system user. And why changing the `repo1-path` and `pg1-path` in the restore command rather than putting it correctly in the config? That means the other commands (like archive-get which will be needed for PG restore_command) will also need those settings.

Checking the `/var/lib/pgsql/test3/recovery.conf` file generated by the restore command might be a good idea too.

> How do you think, should I change pgrbackrest.conf so it knows about a standby?

Did you read the [user-guide](https://pgbackrest.org/user-guide.html#standby-backup) or [this](https://pgstef.github.io/2020/11/21/combining_pgbackrest_dedicated_repository_host_and_streaming_replication.html) old blog post as an example? 

Basically, you add your pgbackrest.conf on the standby server, pointing to your repo-host with only pg1 in the stanza section, and then you add the standby as pg2 option into the stanza section of your repo-host.

**Comment by vvitad** [comment]
> I don't think `repo1-path=/var/lib/pgbackrest` is needed in your primary config.
It's true, it's not there on prod. I should remove it in testing env.

> Did you mounted the production repo or copied/rsynced it? Chmod on your production repo isn't a good idea imo. What you could do is adjust your mount options to map it to the good user:group and only give it RO access.
At first, it was scp. After I've read this https://github.com/pgbackrest/pgbackrest/issues/1971, I mounted it. 
It's all test env, not production. On production I don't have another empty host like test-host here


> Why sudo? The pgbackrest commands should usually run under your PG system user. A
Sudo cause of the mounting, did it for the first time. Didn't know if postgres user will be able to reach there. Checked rn - it can.

> And why changing the `repo1-path` and `pg1-path` in the restore command rather than putting it correctly in the config? That means the other commands (like archive-get which will be needed for PG restore_command) will also need those settings.
I didn't know it will work, so didn't felt the need to change it in conf. 
Also, I was thinking if it's possible to restore on the repo-host, just for testing purposes? In that case if we change pgbackrest.config, won't the stanza break.. we still need it to run daily backups. So for only running a restore once it seems safer to  change the `repo1-path` and `pg1-path`  parameters . 

> Checking the `/var/lib/pgsql/test3/recovery.conf` file generated by the restore command might be a good idea too.
It was like this:
```
# Recovery settings generated by pgBackRest restore on 2023-03-30 XX:XX:XX
restore_command = 'pgbackrest --pg1-path=/var/lib/pgsql/test3 --repo1-path=/mnt/pgbackrest/ --stanza=test3 archive-get %f "%p"'
```
Why?


> Did you read the [user-guide](https://pgbackrest.org/user-guide.html#standby-backup) or [this](https://pgstef.github.io/2020/11/21/combining_pgbackrest_dedicated_repository_host_and_streaming_replication.html) old blog post as an example?
Yep, I couldn't get past the recovery-option in your post, because we use patroni for replicas on production. Is it okay to combine it? 

I've also read your post about patroni&pgbackrest. When you add restore_command, I am confused, because we don't have restore_command in PG 12+. So is it needed on PG < 12 or on PG 12+ too?
>>recovery-option=primary_conninfo=host=pg1-srv user=replic_user

> Basically, you add your pgbackrest.conf on the standby server, pointing to your repo-host with only pg1 in the stanza section, and then you add the standby as pg2 option into the stanza section of your repo-host.
So on prod I have pg-primary, pg-standby1, pg-standby2, repo-host. Replicas were created through patroni (not by me, and I feel afraid coming close to patroni). Pgbackrest know only about pg-primary.
If I copy my pgbackrest.conf from pg-primary to pg-standby1 and 2 servers, with the only difference in pg1-path? It should point on the server where config lies? And only repo-host should have all of them in  pgbackrest.conf?
Do I still need the recovery-option as in your post?

**Comment by pgstef** [comment]
> Also, I was thinking if it's possible to restore on the repo-host, just for testing purposes? In that case if we change pgbackrest.config, won't the stanza break.. we still need it to run daily backups. So for only running a restore once it seems safer to change the `repo1-path` and `pg1-path` parameters .

What do you mean by stanza break? You can manually run a restore with `--reset-pg1-host --archive_mode=off` to test a local restore on a repo host if I remember correctly.

> When you add restore_command, I am confused, because we don't have restore_command in PG 12+. So is it needed on PG < 12 or on PG 12+ too?

`restore_command` is a PG setting that exists since a very very very long time ago. So that would work with 9.6 if that version is supported by Patroni obviously.

> If I copy my pgbackrest.conf from pg-primary to pg-standby1 and 2 servers, with the only difference in pg1-path? It should point on the server where config lies? And only repo-host should have all of them in pgbackrest.conf?

Why would you change the pg1-path on the db nodes? The pg_data should be at the same place...
So the db nodes pgbackrest processes don't need to know anything about the other. And your repo host would need to know about all the db nodes (pg1-, pg2-, pg3- options in the stanza section), so it can determine which one is the primary automatically.

> Do I still need the recovery-option as in your post?

No, that one would be handled by Patroni.

pgBackRest and Patroni does blend pretty well, there are lots of examples / talks about it. At this point, I kinda feel you'd need some training on Patroni and pgBackRest plus some consulting, more than filling GH issues :thinking:

**Comment by dwsteele** [comment]
Agreed -- really none of this is a pgBackRest issue. I suggest you consult the user guide and the links Stefan has provided and get up to speed on PostgreSQL backup and restore.

**Comment by vvitad** [comment]
> What do you mean by stanza break? You can manually run a restore with `--reset-pg1-host --archive_mode=off` to test a local restore on a repo host if I remember correctly.

I found one short mention about it here https://pgbackrest.org/command.html  it's really not enough
 
> `restore_command` is a PG setting that exists since a very very very long time ago. So that would work with 9.6 if that version is supported by Patroni obviously.
It was last mentioned in PG11 documentation. I am asking if I need this in PG12-15.

> Why would you change the pg1-path on the db nodes? 
Because I'm dumb and I meant pg1-host. But I guess we don't need pg1-host in any configs, except the repo?

> The pg_data should be at the same place... So the db nodes pgbackrest processes don't need to know anything about the other. And your repo host would need to know about all the db nodes (pg1-, pg2-, pg3- options in the stanza section), so it can determine which one is the primary automatically.

That's what i meant.
 
Thanks a lot for the consult, I hope it helps someone, cause it definitely helped me. 
 🤔
